File: python-bindings/tests/unpacked_scan_conformance.py
"""Native loose-file scanning with complete read-only filesystem observations."""

import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from file_operations_conformance import _files, _owned_path
from scan_game_conformance import _directories


def _issues(value: Any, root: Path) -> dict[str, list[str]]:
    """Project every native set and preserve DDS paths separately from issue counts."""
    fields = {
        "animation": value.animdata,
        "formats": value.tex_frmt,
        "sounds": value.snd_frmt,
        "scripts": value.xse_file,
        "previs": value.previs,
    }
    count = sum(len(values) for values in fields.values())
    if value.total_count() != count or value.has_issues() is not bool(count):
        raise ValueError("unpacked summary counted DDS inventory as issues")
    result = {
        key: sorted(text.replace("\\", "/") for text in values)
        for key, values in fields.items()
    }
    result["dds"] = sorted(
        path.relative_to(root).as_posix() for path in value.dds_files
    )
    return result


def observe_unpacked_scan(fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Execute public object and convenience scanners over identical owned files."""
    import classic_scangame as native

    with tempfile.TemporaryDirectory(prefix="classic-unpacked-") as directory:
        root = Path(directory)
        for path in fixture["directories"]:
            _owned_path(root, path).mkdir(parents=True, exist_ok=True)
        for path, content in fixture["files"].items():
            target = _owned_path(root, path)
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(content.encode("utf-8"))
        result = {"beforeFiles": _files(root), "beforeDirectories": _directories(root)}
        scanner = native.UnpackedScanner()
        result["issues"] = _issues(
            scanner.scan_directory(root, fixture["scripts"]), root
        )
        if (
            _issues(native.scan_unpacked_files(root, fixture["scripts"]), root)
            != result["issues"]
        ):
            raise ValueError("unpacked convenience scan disagrees with object scan")
        result.update(files=_files(root), directories=_directories(root))
        return result

File: node-bindings/classic-node/__test__/unpacked_scan_conformance.ts
import { mkdir,mkdtemp,readdir,rm,writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname,join,relative } from "node:path";
import { JsUnpackedScanner,scanUnpackedFiles } from "../index.js";
import { inventory } from "./file_generation_conformance.js";

/** Inventory empty directories too so a read-only scan cannot hide filesystem mutations. */
async function directories(root:string,prefix=""):Promise<string[]>{
  const result:string[]=[];for(const entry of await readdir(join(root,prefix),{withFileTypes:true})){const path=prefix?`${prefix}/${entry.name}`:entry.name;if(entry.isDirectory()){result.push(path,...await directories(root,path));}}return result.sort();
}

/** Project actual sets deterministically; DDS paths are inventory, not issue counts. */
function issues(value:Record<string,any>,root:string):Record<string,any>{
  const fields={animation:value.animdata,formats:value.texFrmt,sounds:value.sndFrmt,scripts:value.xseFile,previs:value.previs};
  return {...Object.fromEntries(Object.entries(fields).map(([name,values])=>[name,(values as string[]).map(value=>value.replaceAll("\\","/")).sort()])),dds:value.ddsFiles.map((path:string)=>relative(root,path).replaceAll("\\","/")).sort()};
}

/** Exercise the native class and free-function scanner while retaining all input bytes. */
export async function observeUnpackedScan(fixture:Record<string,any>):Promise<Record<string,any>>{
  const root=await mkdtemp(join(tmpdir(),"classic-unpacked-node-"));
  const target=(path:string)=>{if(path.includes(":")||path.includes("\\")||path.split("/").some(part=>["", ".", ".."].includes(part)))throw new Error("invalid unpacked path");return join(root,path);};
  try{
    for(const path of fixture.directories)await mkdir(target(path),{recursive:true});
    for(const [path,content] of Object.entries(fixture.files)){const destination=target(path);await mkdir(dirname(destination),{recursive:true});await writeFile(destination,content as string,"utf8");}
    const result:Record<string,any>={beforeFiles:await inventory(root),beforeDirectories:await directories(root)};const scanner=new JsUnpackedScanner();result.issues=issues(scanner.scanDirectory(root,fixture.scripts),root);
    if(JSON.stringify(issues(scanUnpackedFiles(root,fixture.scripts),root))!==JSON.stringify(result.issues))throw new Error("unpacked convenience scan changed results");
    result.files=await inventory(root);result.directories=await directories(root);return result;
  }finally{await rm(root,{recursive:true,force:true});}
}
